examples_tests/20.Megatexture/main.cpp

```diff
@@ -14,42 +14,16 @@
 using namespace irr;
 using namespace core;
 
-using STextureData = asset::ITexturePacker::STextureData;
-
-STextureData getTextureData(const asset::ICPUImage* _img, asset::ICPUTexturePacker* _packer, asset::ISampler::E_TEXTURE_CLAMP _uwrap, asset::ISampler::E_TEXTURE_CLAMP _vwrap, asset::ISampler::E_TEXTURE_BORDER_COLOR _borderColor)
-{
-    const auto& extent = _img->getCreationParameters().extent;
-
-    asset::IImage::SSubresourceRange subres;
-    subres.baseMipLevel = 0u;
-    subres.levelCount = core::findLSB(core::roundDownToPoT<uint32_t>(std::max(extent.width, extent.height))) + 1;
-
-    uint8_t border[4]{};//unused anyway
-    auto pgTabCoords = _packer->pack(_img, subres, _uwrap, _vwrap, _borderColor);
-    return _packer->offsetToTextureData(pgTabCoords, _img, _uwrap, _vwrap);
-}
-
-constexpr const char* GLSL_NONUNIFORM_EXT_OVERRIDE = R"(
-#extension GL_EXT_nonuniform_qualifier  : enable
-
-#ifdef IRR_GL_NV_gpu_shader5
-    #define IRR_GL_EXT_nonuniform_qualifier // TODO: we need to overhaul our GLSL preprocessing system to match what SPIRV-Cross actually does
-    #define nonuniformEXT(X) X // TODO: needed until some form of https://github.com/KhronosGroup/SPIRV-Cross/pull/1328 gets merged into master
-#endif
-
-#ifndef IRR_GL_EXT_nonuniform_qualifier
-    #extension GL_ARB_gpu_shader_int64      : require
-    #extension GL_ARB_shader_ballot         : require
-#endif
-)";
-
-constexpr const char* GLSL_VT_TEXTURES = //also turns off set3 bindings (textures) because they're not needed anymore as we're using VT
+constexpr const char* GLSL_VT_DESCRIPTORS = //also turns off set3 bindings (textures) because they're not needed anymore as we're using VT
 R"(
 #ifndef _NO_UV
-layout (set = 0, binding = 0) uniform usampler2DArray pageTable[3];
-layout (set = 0, binding = 1) uniform sampler2DArray physPgTex[3];
-
 #define VT_COUNT 3
+#include <%s> //irr/builtin/glsl/virtual_texturing/descriptors.glsl/...
+layout (set = 2, binding = 1, std430) readonly restrict buffer LUT
+{
+    uint lut[VT_COUNT];
+} layer2pid;
+
 layout (set = 2, binding = 0, std430) readonly restrict buffer PrecomputedStuffUBO
 {
     uint pgtab_sz_log2[VT_COUNT];
```

Comment:
`pgtab_sz_log2` is constant for all layers

Reply:
because page size is constant for all formats, so is `vtex_sz_rcp`

Reply:
You could of course store `phys_pg_tex_sz_rcp[3][MAX_FORMAT_VIEWS_OF_FORMAT_META_CLASS]` but it would be a bit unwieldy  because you'd need to do:
```glsl
float irr_glsl_VT_getPhysPgTexSzRcp(in uint formatMetaclass, uint in formatID)
{
   return phys_pg_tex_sz_rcp[formatMetaclass][formatID];
}
```

For an easier solution you could put the array in an SSBO together with the LUT and store `phys_pg_tex_sz_rcp[PAGE_TABLE_LAYERS]` 
```glsl
float irr_glsl_VT_getPhysPgTexSzRcp(uint in pageTableLayer)
{
   return phys_pg_tex_sz_rcp[pageTableLayer];
}
```

Reply:
> 
> 
> You could of course store `phys_pg_tex_sz_rcp[3][MAX_FORMAT_VIEWS_OF_FORMAT_META_CLASS]` but it would be a bit unwieldy because you'd need to do:
> 
> ```glsl
> float irr_glsl_VT_getPhysPgTexSzRcp(in uint formatMetaclass, uint in formatID)
> {
>    return phys_pg_tex_sz_rcp[formatMetaclass][formatID];
> }
> ```
> 
> For an easier solution you could put the array in an SSBO together with the LUT and store `phys_pg_tex_sz_rcp[PAGE_TABLE_LAYERS]`
> 
> ```glsl
> float irr_glsl_VT_getPhysPgTexSzRcp(uint in pageTableLayer)
> {
>    return phys_pg_tex_sz_rcp[pageTableLayer];
> }
> ```